Revised basic support for half expressions

delcypher opened the pull request:
This work is a culmination of code both in  #904 and #913. It uses the cast approach to represent half floating point immediates. It's not my preferred approach but it will allow to make progress on to more interesting things in the limited time I have left to work on Halide.

abadams commented on src/runtime/tracing.cpp:
```diff
@@ -149,9 +149,12 @@ WEAK int32_t default_trace(void *user_context, const halide_trace_event *e) {
                         ss << ((float *)(e->value))[i];
+                    } else if (print_bits == 16) {
+                        // TODO: We should also emit a decimal approximation
+                        ss.write_float16_from_bits_as_hex_float( ((uint16_t *)(e->value))[i] );
```
This is intended to be a human-readable number. I doubt many humans can read hex floats. I'd cast to double then print instead. If you're tracing and you want the precise value then you're using the binary format anyway, not this code.

delcypher replied:
Should I throw away ``write_float16_from_bits_as_hex_float`` then?

abadams replied:
(though I guess cast-to-double will need to lift some code elsewhere)

abadams replied:
I suppose so. It can always be retrieved from history if needed.

delcypher replied:
Okay as style point it would be syntatically nice to have a ``float16_t`` type in the run time so I can implement ``inf_f16()`` (and others) so that they return a ``float16_t`` rather than ``uint16_t``. Do you have any preference on how that should look (i.e. a typedef or an actual class that mirrors float16_t in libHalide that we can add operations to as needed?)

abadams replied:
The second option sounds best to me. With a typedef there's a risk of operators not doing what they look like they do.

abadams commented on src/IROperator.h:
```diff
@@ -820,7 +885,12 @@ inline Expr pow(Expr x, Expr y) {
+    }
+    else if (x.type() == Float(16)) {
+         y = cast<float16_t>(y);
+        return Internal::Call::make(Float(16), "pow_f16", {x, y}, Internal::Call::Extern);
```
Do any of these functions exist? I don't see implementations anywhere.

delcypher replied:
They don't. The commit message comments on this.

My idea here is that I don't want arguments to these math functions to be implicitly casted to float32 if they are of type float16. No one is using float16 expressions yet so this won't break any existing uses. As I experiment with backends I can add support as I go as needed.

abadams replied:
Oh, just noticed the commit comment. I guess whether or not these exist might be backend-specific, hrm... should we provide them for all backends, with a slow implementation as the default, or should we error out in the backend if we detect a float16 math intrinsic. I suppose backends that don't support float16 math will already have thrown an error.

abadams replied:
ok, sgtm. Maybe add a comment like "these don't currently exist for any backends yet" somewhere.

delcypher replied:
From a correctness stand point erroring out if the backend doesn't support half (I seem recall @zvookin  mentioning teaching halide to declare what types a target supports) seems like the right thing to do.

However from a convenience stand point if someone has gone to trouble of writing their algorithm half types rewriting for other targets will become a nuisance. Maybe halide should expose some sort of optional mode which allows you to treat ``floatX`` as ``floatY`` for a particular target. E.g. if enabled all float16 expressions will be treated as float32 in the backend. This would also allow you to compile code written using doubles on to targets that don't support doubles without rewriting the algorithm.

abadams replied:
You'll probably want to test your float16-using algorithm on an x86 CPU at some stage. On x86 most likely we'll just do the math as float32 then truncate on store, and this is what AVX2 supports and seems to encourage. You won't get bit-accurate results, but I think this falls within the bounds of our -ffast-math style behavior.

delcypher replied:
How would you tell the backend to do truncation? I recently discovered that the LLVM ``fptrunc`` instruction doesn't actually truncate!

http://lists.llvm.org/pipermail/llvm-dev/2015-August/089504.html

abadams replied:
I was thinking of these: https://software.intel.com/en-us/node/524287

There's probably an llvm intrinsic to get it to emit the corresponding instructions. Pre-avx2 we'd have to think of some clever way to emulate it with vectorizable bit tricks.

delcypher replied:
Well there are these

http://llvm.org/docs/LangRef.html#half-precision-floating-point-intrinsics

but they don't specify a rounding mode. Looking at ``IntrinsicsX86.td`` I see these entires. I think these are the intrinsics for the instructions you mention. They allow you to specify the rounding mode which is cool :) but these seem to be vectorised instructions, your link also mentions scalar intrinsics too.

```
// Half float conversion

let TargetPrefix = "x86" in {  // All intrinsics start with "llvm.x86.".
  def int_x86_vcvtph2ps_128 : GCCBuiltin<"__builtin_ia32_vcvtph2ps">,
              Intrinsic<[llvm_v4f32_ty], [llvm_v8i16_ty], [IntrNoMem]>;
  def int_x86_vcvtph2ps_256 : GCCBuiltin<"__builtin_ia32_vcvtph2ps256">,
              Intrinsic<[llvm_v8f32_ty], [llvm_v8i16_ty], [IntrNoMem]>;
  def int_x86_vcvtps2ph_128 : GCCBuiltin<"__builtin_ia32_vcvtps2ph">,
              Intrinsic<[llvm_v8i16_ty], [llvm_v4f32_ty, llvm_i32_ty],
                        [IntrNoMem]>;
  def int_x86_vcvtps2ph_256 : GCCBuiltin<"__builtin_ia32_vcvtps2ph256">,
              Intrinsic<[llvm_v8i16_ty], [llvm_v8f32_ty, llvm_i32_ty],
                        [IntrNoMem]>;
  def int_x86_avx512_mask_vcvtph2ps_512 : GCCBuiltin<"__builtin_ia32_vcvtph2ps512_mask">,
              Intrinsic<[llvm_v16f32_ty], [llvm_v16i16_ty, llvm_v16f32_ty,
                                           llvm_i16_ty, llvm_i32_ty], [IntrNoMem]>;
  def int_x86_avx512_mask_vcvtps2ph_512 : GCCBuiltin<"__builtin_ia32_vcvtps2ph512_mask">,
              Intrinsic<[llvm_v16i16_ty], [llvm_v16f32_ty, llvm_i32_ty,
                                           llvm_v16i16_ty, llvm_i16_ty], [IntrNoMem]>;
}
```

delcypher commented:
@abadams 
Would you be able to quickly glance at 4415a01674e3bc4ebd546b62f98c17ec7fffe417 . I'm thinking that maybe using a typedef  rather than a struct to represent ``halide_float16_t`` would be better given the pain it causes with the header files. Note the implementation of the conversion functions hasn't been fully tested yet, I still need to write tests for it.

abadams commented:
That seems fine, except that I'm not clear on why the definition is
repeated in HalideRuntime.h and runtime_internal.h. The definitions are the
same, right? It seems like HalideRuntime.h should be enough.

On Fri, Aug 28, 2015 at 7:18 PM, Dan Liew <notifications@github.com> wrote:

> @abadams <https://github.com/abadams>
> Would you be able to quickly glance at 4415a01
> <https://github.com/halide/Halide/commit/4415a01674e3bc4ebd546b62f98c17ec7fffe417>
> . I'm thinking that maybe using a typedef rather than a struct to represent
> halide_float16_t would be better given the pain it causes with the header
> files. Note the implementation of the conversion functions hasn't been
> fully tested yet, I still need to write tests for it.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/halide/Halide/pull/919#issuecomment-135931384>.
>

delcypher commented:
> That seems fine, except that I'm not clear on why the definition is
repeated in HalideRuntime.h and runtime_internal.h. The definitions are the
same, right?

Yes the definitions are the same.

> It seems like HalideRuntime.h should be enough.

Unfortunately no. The ``halide_float16_t`` and the associated functions need to be visible in ``HalideRuntime.h`` so that it can be used publicly but they also need to be declared in ``runtime_internal.h`` because the ``Printer`` uses the type and a function that operates on this type. So this type and the functions need to available internally and externally.

If the type is declared both in ``runtime_internal.h`` and in ``HalideRuntime.h`` it causes compilation problems (e.g. ``cache.cpp``) because the include order always seems to be

```
#include "runtime_internal.h"
#include "HalideRuntime.h"
```

which is why the horrible ``#ifndef COMPLILING_HALIDE_RUNTIME`` is but around the declarations in ``HalideRuntime.h`` so ``halide_float16_t`` doesn't get declared twice when compiling.

I think the right solution here would be to have a ``HalideFloat16_t.h`` header file (with the normal include guard) that's included by both ``HalideRuntime.h`` and ``runtime_internal.h`` that contains the declarations. This would need to be a public header file though, would it be okay to introduce this?

delcypher commented:
I've moved ``halide_float16_t`` and associated runtime functions into their own header file and rebased on to master.

This isn't ready to merge yet though as I still need to write some tests.

abadams commented:
Actually I'd rather not expose an additional header to users. Instead how
about we move the Printer class into its own header file?

On Sat, Aug 29, 2015 at 6:09 PM, Dan Liew <notifications@github.com> wrote:

> I've moved halide_float16_t and associated runtime functions into their
> own header file and rebased on to master.
>
> This isn't ready to merge yet though as I still need to write some tests.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/halide/Halide/pull/919#issuecomment-136068911>.
>

delcypher commented:
> Instead how about we move the Printer class into its own header file?

That will work provided ``runtime_internal.h`` doesn't try to include the printer's header file and every user of the printer manually includes the printer header file after ``HalideRuntime.h`` has been included. Is that your intention?

abadams commented:
Yes, that was my intention.

On Sat, Aug 29, 2015 at 6:38 PM, Dan Liew <notifications@github.com> wrote:

> Instead how about we move the Printer class into its own header file?
>
> That will work provided runtime_internal.h doesn't try to include the
> printer's header file and every user of the printer manually includes the
> printer header file after HalideRuntime.h has been included. Is that your
> intention?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/halide/Halide/pull/919#issuecomment-136070293>.
>

delcypher commented:
Okay I've split the ``Printer`` out into its own header file (``printer.h``).

I've also added infrastructure to the build systems to support what I call "pure runtime tests". These are tests that do not link against libhalide and instead just link against Halide's runtime. The convention is that if a test source file starts with the ``runtime_`` prefix it is treated as a "pure runtime test". Hopefully this isn't too controversial.

I've started writing a test (tests conversion from ``halide_float16_t`` to ``float`` and ``double``) that uses this new build system infrastructure but I've not committed it because it's not finished.

delcypher commented:
This is ready to be looked at again.

Unfortunately it is currently blocked by #924

delcypher commented:
@abadams 
I've ended up using the typedef solution.

> It's a long shot, but you might want to test throwing a
HALIDE_ATTRIBUTE_ALIGN(2) on the uint16_t member or a pragma packed on the
struct.

This didn't fix the issue

> Another (also icky) possibility: Make those functions accept a float16_t by
const ref instead of by value, and add a deleted copy constructor if
COMPILING_HALIDE_RUNTIME is defined to stop anyone from passing it by value
in the runtime itself.

The functions are publicly exposed as C functions so I can't use C++ features here. There is an argument to be made here that they shouldn't be publicly exposed as they are only really public so I can test them (the pure runtime test includes ``HalideRuntime.h`` not ``runtime_internal.h``). I think they're quite useful though so having them public could be useful.

When you have time could you take another look at this PR. I think it's ready to go.

abadams commented on src/Expr.h:
```diff
@@ -184,6 +184,10 @@ struct Expr : public Internal::IRHandle {
     }
 
+    /** Make an expression representing a const 16-bit float as a
+     * cast of a (i.e. a FloatImm).*/
```
Looks like there's a word or two missing in this sentence

delcypher replied:
I'll fix that.

abadams commented on src/runtime/HalideRuntime.h:
```diff
@@ -742,6 +742,24 @@ extern void halide_profiler_reset();
+ * This type represents the bits for the IEEE-754 2008 binary16 floating point
+ * format (also known as "half")
+ */
+typedef uint16_t halide_float16_t;
```
Do you think it would be better if we skipped the typedef entirely, and just made the functions below take a uint16_t (and adjust the comments to explain that the bits will be interpreted as an IEE-754 float16)? That way there's no danger of someone adding two halide_float16_ts together and getting a nasty surprise.

delcypher replied:
Sure I'll do that.

abadams commented:
Sorry if this gets slow to merge. I want to test this on Windows, but the Windows llvm build is currently broken due to what looks like a clang change (http://buildbot.halide-lang.org:8010/builders/win-64-trunk/builds/106). To further complicate things, I fly to Australia tomorrow for paternity leave.

abadams commented:
(I'll still be online in Australia, just at a reduced capacity)

abadams commented:
Actually it looks like the llvm windows breakage is already fixed. If you pull it into a branch in the main repo when it's ready to merge, I can instruct the windows buildbots to test that branch. I don't think I can easily instruct them to test a branch in an alternative repo.

delcypher commented:
@abadams 
I've addressed your comments and I've pushed what's in this branch into [a branch of the same name in the Halide repository](https://github.com/halide/Halide/tree/half_expr_revised). Please let me know if there's more I need to do.

abadams commented:
Windows builds started:

http://buildbot.halide-lang.org:8010/builders/win-32-trunk/builds/81
http://buildbot.halide-lang.org:8010/builders/win-64-trunk/builds/109

delcypher commented:
Both builds failed on ``RemoveDirectory_2`` step. I'm not sure if that really has anything to do with my change.

abadams commented:
Might be because I cancelled an earlier build to let this run. Maybe a process from the earlier build still had that folder locked. Will retry.

abadams commented:
Passed once I changed the scripts to ignore failures to delete directories.
